Mathlib/LinearAlgebra/Basic.lean

```diff
@@ -0,0 +1,2723 @@
+  rfl
+#align linear_equiv.fun_congr_left_id LinearEquiv.funCongrLeft_id
+
+@[simp]
+theorem funCongrLeft_comp (e₁ : m ≃ n) (e₂ : n ≃ p) :
+    funCongrLeft R M (Equiv.trans e₁ e₂) =
+      LinearEquiv.trans (funCongrLeft R M e₂) (funCongrLeft R M e₁) :=
+  rfl
+#align linear_equiv.fun_congr_left_comp LinearEquiv.funCongrLeft_comp
+
+@[simp]
+theorem funCongrLeft_symm (e : m ≃ n) : (funCongrLeft R M e).symm = funCongrLeft R M e.symm :=
+  rfl
+#align linear_equiv.fun_congr_left_symm LinearEquiv.funCongrLeft_symm
+
+end LinearEquiv
+
+end FunLeft
+
+namespace LinearMap
+
+variable [Semiring R] [AddCommMonoid M] [Module R M]
+
+variable (R M)
+
+/-- The group of invertible linear maps from `M` to itself -/
+@[reducible]
+def GeneralLinearGroup :=
+  (M →ₗ[R] M)ˣ
+#align linear_map.general_linear_group LinearMap.GeneralLinearGroup
+
+namespace GeneralLinearGroup
+
+variable {R M}
+
+instance : CoeFun (GeneralLinearGroup R M) fun _ => M → M := by infer_instance
+
+/-- An invertible linear map `f` determines an equivalence from `M` to itself. -/
+def toLinearEquiv (f : GeneralLinearGroup R M) : M ≃ₗ[R] M :=
+  { f.val with
+    invFun := f.inv.toFun
+    left_inv := fun m => show (f.inv * f.val) m = m by erw [f.inv_val] <;> simp
+    right_inv := fun m => show (f.val * f.inv) m = m by erw [f.val_inv] <;> simp }
+#align linear_map.general_linear_group.to_linear_equiv LinearMap.GeneralLinearGroup.toLinearEquiv
+
+/-- An equivalence from `M` to itself determines an invertible linear map. -/
+def ofLinearEquiv (f : M ≃ₗ[R] M) : GeneralLinearGroup R M
+    where
+  val := f
+  inv := (f.symm : M →ₗ[R] M)
+  val_inv := LinearMap.ext fun _ => f.apply_symm_apply _
+  inv_val := LinearMap.ext fun _ => f.symm_apply_apply _
+#align linear_map.general_linear_group.of_linear_equiv LinearMap.GeneralLinearGroup.ofLinearEquiv
+
+variable (R M)
+
+/-- The general linear group on `R` and `M` is multiplicatively equivalent to the type of linear
+equivalences between `M` and itself. -/
+def generalLinearEquiv : GeneralLinearGroup R M ≃* M ≃ₗ[R] M
+    where
+  toFun := toLinearEquiv
+  invFun := ofLinearEquiv
+  left_inv f := by
+    ext
+    rfl
+  right_inv f := by
+    ext
+    rfl
+  map_mul' x y := by
+    ext
+    rfl
+#align linear_map.general_linear_group.general_linear_equiv LinearMap.GeneralLinearGroup.generalLinearEquiv
+
+@[simp]
+theorem generalLinearEquiv_to_linearMap (f : GeneralLinearGroup R M) :
+    (generalLinearEquiv R M f : M →ₗ[R] M) = f := by
+  ext
+  rfl
+#align linear_map.general_linear_group.general_linear_equiv_to_linear_map LinearMap.GeneralLinearGroup.generalLinearEquiv_to_linearMap
+
+@[simp]
+theorem coeFn_generalLinearEquiv (f : GeneralLinearGroup R M) :
+    ⇑(generalLinearEquiv R M f) = (f : M → M) :=
+  rfl
+#align linear_map.general_linear_group.coe_fn_general_linear_equiv LinearMap.GeneralLinearGroup.coeFn_generalLinearEquiv
+
+end GeneralLinearGroup
+
+end LinearMap
```

Comment:
Note that we can remove this if #18345 goes through first.

Reply:
That PR is merged, so this _could_ be deleted

Reply:
done in [4bfe92c](https://github.com/leanprover-community/mathlib4/pull/1979/commits/4bfe92c44242d4272277abc381c99dcd20074989)